---
title: "OPTmatch"
output: html_notebook
---

Loaded all functions in for OPT match. 
```{r}
library(Signac)
library(Seurat)
library(GenomeInfoDb)
library(EnsDb.Hsapiens.v86)
library(ggplot2)
library(patchwork)
library(ComplexHeatmap)
library(patchwork)
library(ggrepel)

library(ggplot2)
library(ggrepel)
library(reshape2)

library(ComplexHeatmap)
library(circlize)
library(networkD3)
library(GGally)
library(igraph)
library(network)
library(optmatch)
library(Matrix)
library(FNN)
library(dplyr)

library(ggrastr)
library(BuenColors)
library(stringr)
library(SEtools)
library(SummarizedExperiment)
set.seed(1234)





```

```{r}
setwd("./optmatch/")
source("./utils.R")
source("./optMatching_functions.R")

#We are using the L2 normalized CCA embeddings of our RNA and ATAC cells
CCA_L2_GAMS<-readRDS(file="./Coembed_RNA_ATAC_CCA_GAMS.rds")

#read in integrated object
coembed<-readRDS(file="./Coembed_RNA_ATAC_GAM_int.rds")


#first will have tro do smart up Sample
UMAP<-Embeddings(coembed,reduction = "umap")
```

We use the L2 norm PCA of the coembeded dataset and do the analysis of WT, MUT, and CTR cell separately. 

```{r}
#do analysis for WT, MUT, and CTR separately
coembed$IDH_type<-as.factor(coembed$IDH_type)
levels(coembed$IDH_type)<-c("CTR","MUT","CTR","WT")
PCA_L2_GAMS_WT<-coembed[,coembed$IDH_type =="WT"]
PCA_L2_GAMS_MUT<-coembed[,coembed$IDH_type =="MUT"]
PCA_L2_GAMS_CTR<-coembed[,coembed$IDH_type =="CTR"]

#WT modalities
RNA_pos_WT<-PCA_L2_GAMS_WT$modality=="RNA"
ATAC_pos_WT<-PCA_L2_GAMS_WT$modality=="ATAC"

#WT PCs
All_pcs_WT<-Embeddings(PCA_L2_GAMS_WT,reduction = "pca")
RNApcs_WT<-All_pcs_WT[RNA_pos_WT,]
ATACpcs_WT<-All_pcs_WT[ATAC_pos_WT,]

#Same for MUT
RNA_pos_MUT<-PCA_L2_GAMS_MUT$modality=="RNA"
ATAC_pos_MUT<-PCA_L2_GAMS_MUT$modality=="ATAC"
All_pcs_MUT<-Embeddings(PCA_L2_GAMS_MUT,reduction = "pca")
RNApcs_MUT<-All_pcs_MUT[RNA_pos_MUT,]
ATACpcs_MUT<-All_pcs_MUT[ATAC_pos_MUT,]

#and for CTR cells
RNA_pos_CTR<-PCA_L2_GAMS_CTR$modality=="RNA"
ATAC_pos_CTR<-PCA_L2_GAMS_CTR$modality=="ATAC"
All_pcs_CTR<-Embeddings(PCA_L2_GAMS_CTR,reduction = "pca")
RNApcs_CTR<-All_pcs_CTR[RNA_pos_CTR,]
ATACpcs_CTR<-All_pcs_CTR[ATAC_pos_CTR,]


#ATAC and RNA pcs used for chunked pairing
geo_paired_MUT_pca<-chunk_pair_geo(ATACpcs = ATACpcs_MUT,RNApcs = RNApcs_MUT,nCores=10) 
geo_paired_WT_pca<-chunk_pair_geo(ATACpcs = ATACpcs_WT,RNApcs = RNApcs_WT,nCores=10)
geo_paired_CTR_pca<-chunk_pair_geo(ATACpcs = ATACpcs_CTR,RNApcs = RNApcs_CTR,nCores=10) 


#take UMAP from coembeded object
UMAP<-Embeddings(coembed,reduction = "umap")
UMAP_df<-as.data.frame(UMAP)
names(UMAP_df)<-c("UMAP1","UMAP2")

#Renaming for MUT, CTR, and WT
ATAC_names_MUT_pca<-str_split_fixed(geo_paired_MUT_pca$ATAC,pattern = "_que",n = 2)[,1]
RNA_names_MUT_pca<-str_split_fixed(geo_paired_MUT_pca$RNA,pattern = "_ref",n = 2)[,1]

ATAC_names_CTR_pca<-str_split_fixed(geo_paired_CTR_pca$ATAC,pattern = "_que",n = 2)[,1]
RNA_names_CTR_pca<-str_split_fixed(geo_paired_CTR_pca$RNA,pattern = "_ref",n = 2)[,1]

ATAC_names_WT_pca<-str_split_fixed(geo_paired_WT_pca$ATAC,pattern = "_que",n = 2)[,1]
RNA_names_WT_pca<-str_split_fixed(geo_paired_WT_pca$RNA,pattern = "_ref",n = 2)[,1]

#Show the pairing
MUT_pairs_pca<-plotPairs(ATAC = ATAC_names_MUT_pca,RNA = RNA_names_MUT_pca,umap.df =UMAP_df,seed=5,max.show = 300)
CTR_pairs_pca<-plotPairs(ATAC = ATAC_names_CTR_pca,RNA = RNA_names_CTR_pca,umap.df =UMAP_df,seed=5,max.show = 300)
WT_pairs_pca<-plotPairs(ATAC = ATAC_names_WT_pca,RNA = RNA_names_WT_pca,umap.df =UMAP_df,seed=5,max.show = 300)

MUT_pairs_pca<-MUT_pairs_pca+ggtitle("MUT pairs with corr PCA")
WT_pairs_pca<-WT_pairs_pca+ggtitle("WT pairs with corr PCA")
CTR_pairs_pca<-CTR_pairs_pca+ggtitle("CTR pairs with corr PCA")
MUT_pairs_pca+WT_pairs_pca+CTR_pairs_pca

a<-DimPlot(coembed,group.by = "modality",split.by = "type")

b<-WT_pairs_pca+MUT_pairs_pca+CTR_pairs_pca

#write out all of the paired cells
write.table(geo_paired_MUT_pca,"./geo_paired_MUT_pca.txt")
write.table(geo_paired_WT_pca,"./geo_paired_WT_pca.txt")
write.table(geo_paired_CTR_pca,"./geo_paired_CTR_pca.txt")
saveRDS(coembed,file="./Coembed_RNA_ATAC_GAM_int.rds")
```


We then select for the purest Milo populations based on our reference mapping and milo analyis.
```{r}
#look for pure milo pops based on the Milo Differential abundace results

da_results_WT<-read.csv("./WT.table.nhoods.csv")
da_results_MUT<-read.csv("./MUT.table.nhoods.csv")
da_results_ctr<-read.csv("./CTR.table.nhoods.csv")

Nhood_sel_WT<-da_results_WT[da_results_WT$logFC > 2 & da_results_WT$SpatialFDR<0.001,"Nhood"]
Nhood_sel_MUT<-da_results_MUT[da_results_MUT$logFC > 2 & da_results_MUT$SpatialFDR<0.001,"Nhood"]
Nhood_sel_CTR<-da_results_ctr[da_results_ctr$logFC > 2 & da_results_ctr$SpatialFDR<0.001,"Nhood"]

#See which cells are chosen

ABD.hm.integrated_milo<-readRDS("./Milo_all.rds")
#ABD.hm.integrated<-readRDS("")
purer_neigh_WT<-row.names(ABD.hm.integrated_milo@nhoods)[rowSums(ABD.hm.integrated_milo@nhoods[,Nhood_sel_WT])>0]
purer_neigh_MUT<-row.names(ABD.hm.integrated_milo@nhoods)[rowSums(ABD.hm.integrated_milo@nhoods[,Nhood_sel_MUT])>0]
purer_neigh_CTR<-row.names(ABD.hm.integrated_milo@nhoods)[rowSums(ABD.hm.integrated_milo@nhoods[,Nhood_sel_CTR])>0]

WT<-data.frame(cell=purer_neigh_WT,type=rep("WT",times=length(purer_neigh_WT)))
MUT<-data.frame(cell=purer_neigh_MUT,type=rep("MUT",times=length(purer_neigh_MUT)))
CTR<-data.frame(cell=purer_neigh_CTR,type=rep("CTR",times=length(purer_neigh_CTR)))

write.csv(rbind(WT,MUT,CTR),file = "./PURE.alldatasets.cells.csv")

#plot using ABD pops
ABD.hm.integrated<-readRDS("./Seurat_post_milo_ready.int.rds")
ABD.hm.integrated<-ScaleData(ABD.hm.integrated,features = row.names(ABD.hm.integrated))
pa<-DimPlot(ABD.hm.integrated,cells.highlight = purer_neigh_WT)+ggtitle("purest WT")+NoLegend()
pb<-DimPlot(ABD.hm.integrated,cells.highlight = purer_neigh_MUT)+ggtitle("purest MUT")+NoLegend()
pc<-DimPlot(ABD.hm.integrated,cells.highlight = purer_neigh_CTR)+ggtitle("purest CTR")+NoLegend()

```

Intersect with geo paired lists in this block before FigR
```{r}
M_shccell<-intersect(geo_paired_MUT_pca$RNA,purer_neigh_MUT)
W_shccell<-intersect(geo_paired_WT_pca$RNA,purer_neigh_WT)
CTR_shccell<-intersect(geo_paired_CTR_pca$RNA,purer_neigh_CTR)

cells_keep_RNA<-c(M_shccell,W_shccell,CTR_shccell)

Idents(ABD.hm.integrated)<-as.factor(colnames(ABD.hm.integrated) %in% cells_keep_RNA)

ABD.hm.integrated<-subset(ABD.hm.integrated,idents=TRUE)

M_shccell_ATAC<-geo_paired_MUT_pca[geo_paired_MUT_pca$RNA %in% M_shccell,1]
W_shccell_ATAC<-geo_paired_WT_pca[geo_paired_WT_pca$RNA %in% W_shccell,1]
CTR_shccell_ATAC<-geo_paired_CTR_pca[geo_paired_CTR_pca$RNA %in% CTR_shccell,1]

M_shccell_RNA<-geo_paired_MUT_pca[geo_paired_MUT_pca$RNA %in% M_shccell,2]
W_shccell_RNA<-geo_paired_WT_pca[geo_paired_WT_pca$RNA %in% W_shccell,2]
CTR_shccell_RNA<-geo_paired_CTR_pca[geo_paired_CTR_pca$RNA %in% CTR_shccell,2]

cells_keep_ATAC<-c(M_shccell_ATAC,W_shccell_ATAC,CTR_shccell_ATAC)
cells_keep_RNA<-c(M_shccell_RNA,W_shccell_RNA,CTR_shccell_RNA)

#Test the DE difference between PURE vs NON pure pops

Idents(coembed)<-as.factor(colnames(coembed) %in% c(cells_keep_RNA,cells_keep_ATAC))

coembed_sub<-subset(coembed,idents=TRUE)

FindAllMarkers(ABD.hm.integrated)

write.csv(rbind(M_shccell_ATAC,W_shccell_ATAC,CTR_shccell_ATAC),file = "./PURE.cells.csv")


DE_pure<-FindAllMarkers(ABD.hm.integrated,test.use = "MAST",latent.vars = c("percent.ribo","percent.mt"),only.pos = T)

write.csv(DE_pure,file = "./PURE_DE.csv")

Idents(RNA_GAM)<-RNA_GAM$type
DE_non_pure<-FindAllMarkers(RNA_GAM,test.use = "MAST",latent.vars = c("percent.ribo","percent.mt"),only.pos = T)

write.csv(DE_non_pure,file = "./NON_PURE_DE_WT_MUT.csv")

#See what the top results are with DE
top10_aggr<-c()
for (i in c("WT","MUT","NB"))
{

top10<-grep("^(MALAT1|XIST|HSP|MT|HIST|AC006129.2|RP5-998N21.4|HIF1A-AS2|RP11-1000B6.3|NEAT1)", DE_non_pure[DE_non_pure$cluster==i,]$gene, invert = T, value = T)[1:25]
top10_aggr<-c(top10_aggr,top10)
  
}


data<-data.frame(Level=RNA_GAM$cluster_final,type=RNA_GAM$type,cell_id=row.names(RNA_GAM[[]]))
stratified_rna <- data %>%
group_by(type) %>%
sample_n(size=500)

A<-DoHeatmap(RNA_GAM,cells=stratified_rna$cell_id,features = top10_aggr)
RNA_sel<-t(FetchData(RNA_GAM,vars = top10_aggr,slot = "scale.data"))[,stratified_rna$cell_id]
col_fun = colorRamp2(c(-2,0,2), c("#08306b", "#bcbddc" ,"#67000d"))
A<-Heatmap(RNA_sel,cluster_columns = F,cluster_rows = F,show_column_names = F,col=col_fun)

data<-data.frame(type=ATAC_GAM$type,cell_id=row.names(ATAC_GAM[[]]))
stratified <- data %>%
group_by(type) %>%
sample_n(size=500)

#Create aggregate matrices of ATAC, RNA data in pure populations for plotting.

B<-DoHeatmap(ATAC_GAM,cells=stratified$cell_id,features = top10_aggr) + scale_fill_viridis(option="magma")

a<-AverageExpression(ATAC_GAM,features = top10_aggr)

C=Heatmap(t(scale(t(as.matrix(a$RNA)))),cluster_columns = F,cluster_rows = F,col=col_fun)

A+B+C



Idents(ABD.hm.integrated)<-"IDH_simp"
DoHeatmap(object = ABD.hm.integrated,features  = c(head(DE_pure$gene[DE_pure$cluster=="WT"],n=10),head(DE_pure$gene[DE_pure$cluster=="MUT"],n=10),head(DE_pure$gene[DE_pure$cluster=="CTR"],n=10)))

write.csv(cells_keep_ATAC,"./cells_keep_ATAC.csv")
write.csv(cells_keep_RNA,"./cells_keep_RNA.csv")

```


In this section we prep for DORC calling for PURE populations. We tested this both in the context of Seurat and by using FigR.
```{r}
#from here insert pairing form the separate PCA analysis and pure cells

source("./stimATAC_analyses_code/R/DORC_functions.R")
source("./stimATAC_analyses_code/R/utils.R")

#create hg38 refseq
hg38ref<-read.table("./GAM_seq/data/refseq_hg38_UCSC_TSS_unique.bed",header = F)
names(hg38ref)<-c("seqnames","start","end","transcript_id","gene_name","extra","strand")
hg38TSSRanges<-makeGRangesFromDataFrame(hg38ref,keep.extra.columns = TRUE)
save(hg38TSSRanges,file="./GAM_seq/data/hg38TSSRanges.rData")
load("./GAM_seq/data/hg38TSSRanges.rData")



# ---------------------------------------------------------------------------------- Load data

unintegrated_b1_b2_RNA<-readRDS("./RNA_GAM_sel.rds")


#subset to pure in RNA 
unintegrated_b1_b2_RNA_temp<-unintegrated_b1_b2_RNA[,unique(sort(cells_keep_RNA$x))]
Idents(unintegrated_b1_b2_RNA_temp)<-"IDH_type"
what_features_to_keep<-FindAllMarkers(unintegrated_b1_b2_RNA_temp,test.use = "MAST",min.pct = 0.1,latent.vars = c("percent.ribo","percent.mt"))
what_features_to_keep<-unique(sort(what_features_to_keep$gene[what_features_to_keep$p_val_adj<0.05]))
what_features_to_keep<-what_features_to_keep[-grep(what_features_to_keep,pattern = "^MT")]
what_features_to_keep<-what_features_to_keep[-grep(what_features_to_keep,pattern = "^RP")]
what_features_to_keep<-what_features_to_keep[!what_features_to_keep=="MALAT1"]

rnaMat <- GetAssayData(unintegrated_b1_b2_RNA,slot = "data")[what_features_to_keep,cells_keep_RNA$x]

  
#subset to pure in ATAC and subset to peaks near variable genes
unintegrated_b1_b2_ATAC<-readRDS(file="./ATAC_gene_activity/ATAC_final_subset_2_0805_gene_activity.rds")
DefaultAssay(unintegrated_b1_b2_ATAC)<-"merge.peaks"
unintegrated_b1_b2_ATAC
unintegrated_b1_b2_ATAC[["peaks"]]<-NULL
#filter extra chromosomes
rowRanges = StringToGRanges(row.names(unintegrated_b1_b2_ATAC))
rowRanges<-rowRanges[seqnames(rowRanges) %in% c(paste0("chr",1:22),"chrX","chrY")]

annot<-Annotation(unintegrated_b1_b2_ATAC)
annot<-annot[annot$gene_name %in% what_features_to_keep,]
all_ranges<-reduce(Extend(annot,upstream = 500000,downstream = 500000))
rowRanges<-rowRanges[unique(subjectHits(findOverlaps(all_ranges,rowRanges)))]

unintegrated_b1_b2_ATAC[["merged"]]<-NULL
unintegrated_b1_b2_ATAC[["RNA"]]<-NULL

features_ATAC<-GRangesToString(sort(rowRanges))
counts=GetAssayData(unintegrated_b1_b2_ATAC,slot = "counts",assay = "merge.peaks")[,c(cells_keep_ATAC$x)]
features_ATAC<-row.names(counts)[rowSums(counts)>50]
rowRanges<-StringToGRanges(features_ATAC)
counts=GetAssayData(unintegrated_b1_b2_ATAC,slot = "counts",assay = "merge.peaks")[features_ATAC,c(cells_keep_ATAC$x)]

SE.filt <- SummarizedExperiment(list(counts=counts),rowRanges = sort(rowRanges))

#save ATA
saveRDS(SE.filt,"./All_peaks_near_variable_genes.rds")
saveRDS(rnaMat,"./All_variable_genes.rds")
```

Create a faux multiomic dataset with the pure pairings for calling DORCS for Seurat.
```{r}
#RNA cells and feature sel
unintegrated_b1_b2_RNA<-readRDS("./RNA_GAM_sel.rds")
unintegrated_b1_b2_RNA

unintegrated_b1_b2_RNA_temp<-unintegrated_b1_b2_RNA[,unique(sort(cells_keep_RNA$x))]
Idents(unintegrated_b1_b2_RNA_temp)<-"IDH_type"
what_features_to_keep<-FindAllMarkers(unintegrated_b1_b2_RNA_temp,test.use = "MAST",min.pct = 0.1,latent.vars = c("percent.ribo","percent.mt"))
what_features_to_keep<-unique(sort(what_features_to_keep$gene[what_features_to_keep$p_val_adj<0.05]))
what_features_to_keep<-what_features_to_keep[-grep(what_features_to_keep,pattern = "^MT")]
what_features_to_keep<-what_features_to_keep[-grep(what_features_to_keep,pattern = "^RP")]
what_features_to_keep<-what_features_to_keep[!what_features_to_keep=="MALAT1"]
rnaMat <- GetAssayData(unintegrated_b1_b2_RNA,slot = "counts")[what_features_to_keep,cells_keep_RNA$x]

#ATAC cell and feature selection
unintegrated_b1_b2_ATAC<-readRDS(file="./ATAC_gene_activity/ATAC_final_subset_2_0805_gene_activity.rds")
DefaultAssay(unintegrated_b1_b2_ATAC)<-"merge.peaks"
unintegrated_b1_b2_ATAC
unintegrated_b1_b2_ATAC[["peaks"]]<-NULL
#filter extra chromosomes
rowRanges = StringToGRanges(row.names(unintegrated_b1_b2_ATAC))
rowRanges<-rowRanges[seqnames(rowRanges) %in% c(paste0("chr",1:22),"chrX","chrY")]

annot<-Annotation(unintegrated_b1_b2_ATAC)
annot<-annot[annot$gene_name %in% what_features_to_keep,]

all_ranges<-reduce(Extend(annot,upstream = 500000,downstream = 500000))

rowRanges<-rowRanges[unique(subjectHits(findOverlaps(all_ranges,rowRanges)))]
unintegrated_b1_b2_ATAC[["merged"]]<-NULL
unintegrated_b1_b2_ATAC[["RNA"]]<-NULL
features_ATAC<-GRangesToString(sort(rowRanges))
counts=GetAssayData(unintegrated_b1_b2_ATAC,slot = "counts",assay = "merge.peaks")[,c(cells_keep_ATAC$x)]
features_ATAC<-row.names(counts)[rowSums(counts)>50]
rowRanges<-StringToGRanges(features_ATAC)
rowRanges<-rowRanges[seqnames(rowRanges) %in% c(paste0("chr",1:22),"chrX","chrY")]
features_ATAC<-GRangesToString(rowRanges)
ATAC_counts=GetAssayData(unintegrated_b1_b2_ATAC,slot = "counts",assay = "merge.peaks")[features_ATAC,c(cells_keep_ATAC$x)]

meta_atac<-data.frame(patient_atac=unintegrated_b1_b2_ATAC$patient,type_atac=unintegrated_b1_b2_ATAC$type)
row.names(meta_atac)<-row.names(unintegrated_b1_b2_ATAC[[]])

meta_atac<-meta_atac[colnames(ATAC_counts),]

meta_rna<-data.frame(patient_rna=unintegrated_b1_b2_RNA$patient,type_rna=unintegrated_b1_b2_RNA$IDH_type)
row.names(meta_rna)<-row.names(unintegrated_b1_b2_RNA[[]])
meta_rna<-meta_rna[colnames(rnaMat),]

meta_comb<-cbind(meta_atac,meta_rna)
row.names(meta_rna)<-paste("Cell",1:4932,sep = "_")
row.names(meta_atac)<-paste("Cell",1:4932,sep = "_")
meta_comb<-cbind(meta_atac,meta_rna)

colnames(rnaMat)<-paste("Cell",1:4932,sep = "_")
colnames(ATAC_counts)<-paste("Cell",1:4932,sep = "_")

Paired_pure<-CreateSeuratObject(rnaMat,meta.data = meta_comb)
Paired_pure[["ATAC"]]<-CreateChromatinAssay(ATAC_counts,genome = "hg38",ranges = StringToGRanges(row.names(ATAC_counts)))

DefaultAssay(Paired_pure) <- "RNA"
Paired_pure <- SCTransform(Paired_pure)
Paired_pure <- RunPCA(Paired_pure)

DefaultAssay(Paired_pure) <- "ATAC"
Paired_pure <- FindTopFeatures(Paired_pure, min.cutoff = 5)
Paired_pure <- RunTFIDF(Paired_pure)
Paired_pure <- RunSVD(Paired_pure)

Paired_pure <- FindMultiModalNeighbors(
  object = Paired_pure,
  reduction.list = list("pca", "lsi"), 
  dims.list = list(1:50, 2:40),
  modality.weight.name = "RNA.weight",
  verbose = TRUE
)

Paired_pure <- RunUMAP(
  object = Paired_pure,
  nn.name = "weighted.nn",
  assay = "RNA",
  verbose = TRUE
)

DimPlot(Paired_pure, label = TRUE, repel = TRUE, reduction = "umap",group.by = "type_rna") + NoLegend()
DimPlot(Paired_pure, label = TRUE, repel = TRUE, reduction = "umap",group.by = "patient_rna") + NoLegend()
DefaultAssay(Paired_pure) <- "RNA"
genes<-row.names(Paired_pure)
DefaultAssay(Paired_pure) <- "ATAC"

# first compute the GC content for each peak
Paired_pure <- RegionStats(Paired_pure, genome = BSgenome.Hsapiens.UCSC.hg38)

# get gene annotations for hg38
annotation <- GetGRangesFromEnsDb(ensdb = EnsDb.Hsapiens.v86)
seqlevelsStyle(annotation) <- "UCSC"

Annotation(Paired_pure)<-annotation
```

This part is for calling Peak to gene associations via Seurat/Signac.
```{r}
# link peaks to genes
Paired_pure <- LinkPeaks(
  object = Paired_pure,
  peak.assay = "ATAC",
  expression.assay = "SCT",genes.use = genes
)

saveRDS(Paired_pure,"./Faux_multi.rds")


Idents(Paired_pure)<-as.factor(!duplicated(cells_keep_ATAC$x))
cell_rename<-cells_keep_ATAC$x[!duplicated(cells_keep_ATAC$x)]

Paired_pure_rename<-subset(Paired_pure,idents=TRUE)

names_save<-paste0("combined_",gsub("^_+","",cell_rename))
names_save<-gsub(pattern = "1117pm",replacement = "1117",x = gsub(pattern = "combined_combined_",replacement = "combined_",gsub(pattern = "^combined_ATAC_",replacement = "ATAC_",x=names_save)))

Paired_pure_rename<-RenameCells(Paired_pure_rename,new.names = names_save)

notmatched_names<-data.frame(rename=names_save[!(names_save %in% row.names(all[[]]))][c(8:1108)])
notmatched_names$orig<-str_split_fixed(string = notmatched_names$rename,n = 3,pattern = "_")[,3]
notmatched_names$samp<-str_split_fixed(string = notmatched_names$rename,n = 3,pattern = "_")[,2]

#create fragments as well
fragments<-Fragments(all)

for (i in c(1:19)){Fragments(Paired_pure_rename)<-fragments[[i]]
}

#then removed ones thatg were not in the assay
Fragments(Paired_pure_rename)<-fragments



fragments_nb1 <- CreateFragmentObject(
  path = "./NB/NORMALS/N1/fragments.tsv.gz",cells = notmatched_names$orig[notmatched_names$samp=="NB1"],
  validate.fragments = TRUE
)
ren_NB1<-Cells(fragments_nb1)
names(ren_NB1)<-paste0("ATAC_NB1_",Cells(fragments_nb1))
Cells(fragments_nb1)<-ren_NB1


fragments_nb2 <- CreateFragmentObject(
  path = "./NB/NORMALS/N2/fragments.tsv.gz",cells = notmatched_names$orig[notmatched_names$samp=="NB2"],
  validate.fragments = TRUE
)
ren_NB2<-Cells(fragments_nb2)
names(ren_NB2)<-paste0("ATAC_NB2_",Cells(fragments_nb2))
Cells(fragments_nb2)<-ren_NB2

fragments[[18]]<-fragments_nb1
fragments[[19]]<-fragments_nb1
Fragments(Paired_pure_rename)<-fragments

all<-readRDS("./ATAC_all_unint.rds")
saveRDS(Paired_pure_rename,"./Faux_multi_unique.rds")

#what if we use this faux to do gene_peak associations

Paired_pure_rename<-readRDS("./Faux_multi_unique.rds")

small_color_map <- c('#FBDF88','#00B2CA','#DF5368')
DimPlot(Paired_pure_rename,cols = small_color_map)

CoveragePlot(Paired_pure_rename,cols = small_color_map,region = "VIM")

links<-Links(Paired_pure)

sig_links<-links[links$pvalue<0.05,]
```

Same for DORC calling before FigR
```{r}
# ---------------------------------------------------------------------------------- Call gene-peak associations

# Use 8 cores here
cisCor_all <- runGenePeakcorr(ATAC.se = SE.filt,
                           RNAmat = rnaMat,
                           genome = "hg38",
                           windowPadSize = 50000,
                           nCores = 4,
                           n_bg = 100,
                           p.cut = NULL)

 
write.table(cisCor_all,file="./CisCor.WT_MUT_CTR.all_old_met.txt",quote = F,col.names = T,row.names = F)

# Filter associations using correlation p-value 
cisCor.filt <- cisCor_all %>% dplyr::filter(pvalZ <= 0.05)


# ---------------------------------------------------------------------------------- Determine DORC genes

# Make J plot to rank genes by # significant peaks
dorcGenes <- dorcJplot(dorcTab = cisCor.filt,
                       cutoff = 5,
                       returnGeneList = TRUE)

write.csv(dorcGenes,file="./Gene_networks_with_more_than_3_peaks_sep_analysis.csv")
write.csv(dorcGenes,file="./Gene_networks_with_more_than_5 peaks_sep_analysis_all.csv")
write.csv(dorcGenes,file="./Gene_networks_with_more_than_7 peaks_sep_analysis_old.csv")

```


Running FigR on Pure populations. Here we tried several option. 1) Run on everything 2) Run on only DE genes 3) Run on all TFs 4) Run on TFs in Dorcs 5) Run on WT, MUT, CTR separately. In the end we used 1) and 3).
```{r}
# Script to run core FigR function on stimulation data using new human motif database

cells_keep_ATAC<-read.csv("./cells_keep_ATAC.csv")
cells_keep_RNA<-read.csv("./cells_keep_RNA.csv")

cisCor_all
cisCor_all<-data.frame(Peak=match(links$peak,row.names(Paired_pure_rename)),Gene=links$gene,rObs=links$score,pvalZ=links$pvalue)

#ATAC
unintegrated_b1_b2_ATAC<-readRDS(file="./ATAC_gene_activity/ATAC_final_subset_2_0805_gene_activity.rds")

#ATAC
Paired_pure_rename

DefaultAssay(unintegrated_b1_b2_ATAC)<-"merge.peaks"
#DefaultAssay(Paired_pure_rename)<-"ATAC"


counts=GetAssayData(Paired_pure_rename,slot = "counts",assay = "ATAC")
SE.filt <- SummarizedExperiment(list(counts=counts),rowRanges = StringToGRanges(row.names(counts)))


source("./stimATAC_analyses_code/R/FigR_functions.R")
source("./stimATAC_analyses_code/R/utils.R")
# ---------------------------------------------------------------------------------- Load DORC calls

# DORC calls, all cells 
sigGP <- cisCor.filt

# # Only work in space of DORC genes, choose ones the have more than 3 peaks
sigGP <- sigGP[sigGP$Gene %in% dorcGenes,]

dim(sigGP)

# Normalized counts from SEs subset to only pairs
unintegrated_b1_b2_RNA<-readRDS(file="./RNA_GAM_sel.rds")
rnaMat <- GetAssayData(unintegrated_b1_b2_RNA,slot = "data")[,cells_keep_RNA$x]


#normalize aggregate accessibility

row.names(unintegrated_b1_b2_ATAC)[cisCor.filt$Peak]
row.names(unintegrated_b1_b2_ATAC)[28]

cisCor.filt<-cisCor.filt[order(cisCor.filt$Peak),]
counts_norm=GetAssayData(unintegrated_b1_b2_ATAC,slot = "data",assay = "merge.peaks")[cisCor.filt$Peak,cells_keep_ATAC$x]

agg_counts_norm<-aggregate(counts_norm,by=list(as.factor(cisCor.filt$Gene)),FUN = mean)
row.names(agg_counts_norm)<-agg_counts_norm$Group.1
agg_counts_norm$Group.1<-NULL

DORC.SE <- SummarizedExperiment(list(counts=agg_counts_norm))[,cells_keep_ATAC$x]

dorcMat <- assays(DORC.SE)$counts 
dim(dorcMat)

# ---------------------------------------------------------------------------------- Smooth data
ATAC<-readRDS(file="./ATAC_gene_activity/ATAC_final_subset_2_0805_gene_activity.rds")

#here probably can use lsi for smoothing
lsi <- Embeddings(ATAC,reduction="lsi")[colnames(DORC.SE),]
#lsi <- Embeddings(Paired_pure_rename,reduction="lsi")[colnames(DORC.SE),]
head(lsi)

stopifnot(all(colnames(DORC.SE) %in% rownames(lsi)))

# Get KNNs
lsi.knn <- FNN::get.knn(lsi,k=20)$nn.index
rownames(lsi.knn) <- colnames(DORC.SE)

library(doParallel)

#here we can decide to use all cells that are paired or the purified ones ~37k all cells, 4932 purified
dim(rnaMat)
dim(SE.filt)
#vs
dim(DORC.SE
    )

#first lets do purified

dorcMat.smoothed <- smoothScoresNN(lsi.knn,dorcMat,nCores = 6)

dorcMat.smoothed.dorc<-dorcMat.smoothed[dorcGenes,]

stopifnot(all.equal(rownames(dorcMat.smoothed.dorc),dorcGenes))

rnaMat.dorc<-rnaMat[dorcGenes,cells_keep_RNA$x]
rownames(rnaMat.dorc) <- gsub(x=rownames(rnaMat.dorc),pattern = "-",replacement = "",fixed = TRUE)


#see which are up vs downreg in RNA

Idents(unintegrated_b1_b2_RNA)<-unintegrated_b1_b2_RNA$IDH_type
pos_markers_MUT<-FindMarkers(unintegrated_b1_b2_RNA,ident.1 = "MUT",only.pos = T)
pos_markers_WT<-FindMarkers(unintegrated_b1_b2_RNA,ident.1 = "WT",only.pos = T)
pos_markers_CTR<-FindMarkers(unintegrated_b1_b2_RNA,ident.1 = "CTR",only.pos = T)


MUT_genes<-intersect(row.names(pos_markers_MUT[pos_markers_MUT$p_val_adj<0.05,]),dorcGenes)
WT_genes<-intersect(row.names(pos_markers_WT[pos_markers_WT$p_val_adj<0.05,]),dorcGenes)
CTR_genes<-intersect(row.names(pos_markers_CTR[pos_markers_CTR$p_val_adj<0.05,]),dorcGenes)

rm(DORC.SE)
gc()

counts_rna<-GetAssayData(unintegrated_b1_b2_RNA,slot = "data")[,cells_keep_RNA$x]
RNA.SE<- SummarizedExperiment(list(counts=counts_rna))

#lsi.knn<-lsi.knn[cells_keep_ATAC$x,]
# Just so that the smoothing function will work, since it checks for matching attributes
rownames(lsi.knn) <- colnames(rnaMat.dorc)

rnaMat.dorc


# Run only on TFs to save time
human_pwms_v3 <- readRDS("./stimATAC_analyses_code/data/cisBP_human_pfms_2021.rds")
rnaMat.smoothed <- smoothScoresNN(NNmat = lsi.knn,TSSmat = rnaMat.dorc,nCores = 6,geneList=intersect(rownames(rnaMat.dorc),names(human_pwms_v3)))
gc()


genes_de<-gsub(x=c(MUT_genes,WT_genes),pattern = "-",replacement = "",fixed = TRUE)

rnaMat.smoothed_de <- smoothScoresNN(NNmat = lsi.knn,TSSmat = rnaMat.dorc,nCores = 6,geneList=genes_de)
rnaMat.tf<-rnaMat
rownames(rnaMat.tf) <- gsub(x=rownames(rnaMat),pattern = "-",replacement = "",fixed = TRUE)
rnaMat.smoothed_tf_only <- smoothScoresNN(NNmat = lsi.knn,TSSmat = rnaMat.tf,nCores = 6,geneList=intersect(rownames(rnaMat.tf),names(human_pwms_v3)))


#annotation 
#smoothed dorc plot
dorc_names<-str_split_fixed(colnames(dorcMat.smoothed.dorc),pattern = "\\...",n=2)[,1]
rna_names<-str_split_fixed(colnames(rnaMat.smoothed_de),pattern = "\\.\\.\\.",n=2)[,1]
#rna_names<-paste0(str_split_fixed(rna_names,pattern = "_[0-9]$",n=2)[,1],"_1")

cellInfo <- data.frame(seuratCluster=as.factor(unintegrated_b1_b2_ATAC$type))
cellInfo2 <- data.frame(seuratCluster=as.factor(unintegrated_b1_b2_ATAC$patient))
cellInfo3 <- data.frame(seuratCluster=as.factor(unintegrated_b1_b2_ATAC$grade))

cellsannot<-cellInfo[match(dorc_names,rownames(cellInfo)),]
cellsannot2<-cellInfo2[match(dorc_names,rownames(cellInfo2)),]
cellsannot3<-cellInfo3[match(dorc_names,rownames(cellInfo3)),]

cellInfob <- data.frame(seuratCluster=as.factor(unintegrated_b1_b2_RNA$type))
cellInfo2b <- data.frame(seuratCluster=as.factor(unintegrated_b1_b2_RNA$patient))
cellInfo3b <- data.frame(seuratCluster=as.factor(unintegrated_b1_b2_RNA$grade))

cellsannotb<-cellInfob[match(rna_names,rownames(cellInfob)),]
cellsannot2b<-cellInfo2b[match(rna_names,rownames(cellInfo2b)),]
cellsannot3b<-cellInfo3b[match(rna_names,rownames(cellInfo3b)),]


ha1 = HeatmapAnnotation(WTvsMUT = as.factor(cellsannot), patient=as.factor(cellsannot2), grade=as.factor(cellsannot3),
    annotation_legend_param = list(
        WTvsMUT = list(
                title = "WTvsMUT",
                at = c("WT","MUT" ),
                labels = c("WT", "MUT")
            ),
        patient = list(
                title = "Patient",
                at = c("0218","0308","0330","0405","0603","1012","1013","1116","1119","117am","117pm","0413","0416","0428","0628"),
                labels = c("0218","0308","0330","0405","0603","1012","1013","1116","1119","117am","117pm","0413","0416","0428","0628")
            ),
        grade = list(
                title = "grade",
                at = c(2,3,4),
                labels = c("2", "3","4")
            )
        
        
        
        ),col = list(WTvsMUT=c("WT"="#cab2d6","MUT"="#6a3d9a"),patient = c("0218"="#543005","0308"="#8c510a","0330"="#bf812d","0405"="#dfc27d","0603"="#f6e8c3","1012"="#f5f5f5","1013"="#c7eae5","1116"="#80cdc1","1119"="#35978f","117am"="#01665e","117pm"="#003c30","0413"="grey20","0416"="grey40","0428"="grey60","0628"="grey80"),grade = c("2"="blue","3"="red","4"="green")))


ha2 = HeatmapAnnotation(WTvsMUT = as.factor(cellsannotb), patient=as.factor(cellsannot2b), grade=as.factor(cellsannot3b),
    annotation_legend_param = list(
        WTvsMUT = list(
                title = "WTvsMUT",
                at = c("WT","MUT" ),
                labels = c("WT", "MUT")
            ),
        patient = list(
                title = "Patient",
                at = c("0218","0308","0330","0405","0603","1012","1013","1116","1119","117am","117pm","0413","0416","0428","0628"),
                labels = c("0218","0308","0330","0405","0603","1012","1013","1116","1119","117am","117pm","0413","0416","0428","0628")
            ),
        grade = list(
                title = "grade",
                at = c(2,3,4),
                labels = c("2", "3","4")
            )
        
        
        
        ),col = list(WTvsMUT=c("WT"="#cab2d6","MUT"="#6a3d9a"),patient = c("0218"="#543005","0308"="#8c510a","0330"="#bf812d","0405"="#dfc27d","0603"="#f6e8c3","1012"="#f5f5f5","1013"="#c7eae5","1116"="#80cdc1","1119"="#35978f","117am"="#01665e","117pm"="#003c30","0413"="grey20","0416"="grey40","0428"="grey60","0628"="grey80"),grade = c("2"="blue","3"="red","4"="green")))

#cividis(3)
library(circlize)
col_fun = colorRamp2(c(-4, 0, 4), c("#a50026","#ffffbf","#313695"))

row_split<-c(rep("MUT gene",times=length(MUT_genes)),rep("WT gene",times=length(WT_genes)))
column_split<-as.factor(cellsannot)

H1<-Heatmap(matrix = t(scale(t(as.matrix(dorcMat.smoothed.dorc))))[c(MUT_genes,WT_genes),],show_row_names = F,show_column_names = F,cluster_columns = F,cluster_rows = F,top_annotation = ha1,name = "Scaled DORC accessibility",raster_by_magick = F,row_split = row_split, column_split =column_split,use_raster=F)

row_splitb<-c(rep("MUT gene",times=length(MUT_genes)),rep("WT gene",times=length(WT_genes)))
column_splitb<-as.factor(cellsannotb)

RNA_all <- FindVariableFeatures(unintegrated_b1_b2_RNA, selection.method = "vst", nfeatures = 20000)
RNA_all <- ScaleData(RNA_all, features = c(MUT_genes,WT_genes),vars.to.regress = "percent.mt",verbose = FALSE)

H3b<-Heatmap(matrix = as.matrix(GetAssayData(RNA_all,slot = "scale.data"))[c(MUT_genes,WT_genes),c(RNA_names_WT_pca,RNA_names_MUT_pca)],show_row_names = F,show_column_names = F,cluster_columns = F,cluster_rows = T,top_annotation = ha2,name = "Scaled RNA",raster_by_magick = F,row_split = row_splitb, column_split =column_splitb,use_raster=F)

H2<-Heatmap(matrix = as.matrix(rnaMat.dorc)[genes_de,],show_row_names = F,show_column_names = F,cluster_columns = F,cluster_rows = T)

#pearson corr


A<-as.matrix(rnaMat.smoothed_de)[genes_de,]
B<-as.matrix(dorcMat.smoothed.dorc)[c(MUT_genes,WT_genes),]


dim(B
    )


rnaMat.smoothed_de_and_TF<-rnaMat.smoothed[intersect(row.names(rnaMat.smoothed),genes_de),]


# ---------------------------------------------------------------------------------- Run FigR on SE filt and RNA (pure) select TF

stim_FigR <- runFigR(ATAC.se = SE.filt,
                           dorcK = 30,
                           dorcTab = sigGP,
                           genome = "hg38", 
                           dorcMat = dorcMat.smoothed.dorc,
                           rnaMat = rnaMat.smoothed,
                           n_bg = 50,
                           nCores = 8)

write.table(stim_FigR,file = "./FigR_res_tf_with_DORCS_separate_all.txt",sep = "\t",quote = F,row.names = T,col.names = T)




#now with all TFs

stim_FigR.all.tf <- runFigR(ATAC.se = SE.filt,
                           dorcK = 30,
                           dorcTab = sigGP,
                           genome = "hg38", 
                           dorcMat = dorcMat.smoothed.dorc,
                           rnaMat = rnaMat.smoothed_tf_only,
                           n_bg = 50,
                           nCores = 8)

write.table(stim_FigR.all.tf,file = "./FigR_res_all_tf_DORCS_separate_all.txt",sep = "\t",quote = F,row.names = T,col.names = T)


#only with WT cells 

stim_FigR_WT <- runFigR(ATAC.se = SE.filt,
                           dorcK = 30,
                           dorcTab = sigGP,
                           genome = "hg38", 
                           dorcMat = dorcMat.smoothed.dorc[,1:length(ATAC_names_WT_pca)],
                           rnaMat = rnaMat.smoothed[,1:length(RNA_names_WT_pca)],
                           n_bg = 50,
                           nCores = 8)

write.table(stim_FigR_WT,file = "./FigR_res_only_wt_cells_DORCS_separate_5cutoff.txt",sep = "\t",quote = F,row.names = T,col.names = T)


#only with MUT cells
stim_FigR_MUT <- runFigR(ATAC.se = SE.filt,
                           dorcK = 30,
                           dorcTab = sigGP,
                           genome = "hg38", 
                           dorcMat = dorcMat.smoothed.dorc[,length(ATAC_names_WT_pca)+1:length(ATAC_names_MUT_pca)],
                           rnaMat = rnaMat.smoothed[,length(RNA_names_WT_pca)+1:length(RNA_names_MUT_pca)],
                           n_bg = 50,
                           nCores = 8)

write.table(stim_FigR_MUT,file = "./FigR_res_only_mut_cells_DORCS_separate_5cutoff.txt",sep = "\t",quote = F,row.names = T,col.names = T)



#now with only DE genes
stim_FigR_de <- runFigR(ATAC.se = SE.filt,
                           dorcK = 30,
                           dorcTab = sigGP,
                           genome = "hg38", 
                           dorcMat = dorcMat.smoothed.dorc[c(MUT_genes,WT_genes),],
                           rnaMat = rnaMat.smoothed_de_and_TF,
                           n_bg = 50,
                           nCores = 8)



write.table(stim_FigR_de,file = "./FigR_res_only_DE_cells_DORCS_separate.txt",sep = "\t",quote = F,row.names = T,col.names = T)


#only DE and MUT
stim_FigR_de_MUT <- runFigR(ATAC.se = SE.filt,
                           dorcK = 30,
                           dorcTab = sigGP,
                           genome = "hg38", 
                           dorcMat = dorcMat.smoothed.dorc[c(MUT_genes),],
                           rnaMat = rnaMat.smoothed,
                           n_bg = 50,
                           nCores = 8)

write.table(stim_FigR_de_MUT,file = "./FigR_res_only_DE_MUT_DORCS_separate.txt",sep = "\t",quote = F,row.names = T,col.names = T)


#only DE and WT
stim_FigR_de_WT <- runFigR(ATAC.se = SE.filt,
                           dorcK = 30,
                           dorcTab = sigGP,
                           genome = "hg38", 
                           dorcMat = dorcMat.smoothed.dorc[c(WT_genes),],
                           rnaMat = rnaMat.smoothed,
                           n_bg = 50,
                           nCores = 8)


write.table(stim_FigR_de_WT,file = "./FigR_res_only_DE_WT_DORCS_separate.txt",sep = "\t",quote = F,row.names = T,col.names = T)


library(ggplot2)
library(ggrastr)
library(BuenColors)
library(ggrepel)
set.seed(42)

dat <- subset(stim_FigR, (0.5 < Corr.log10P & Enrichment.log10P > 0.5 ) | ( -0.5 > Corr.log10P & Enrichment.log10P > 0.5))

gAll <- ggplot(stim_FigR,aes(Corr.log10P,Enrichment.log10P,color=Score)) + 
  geom_point_rast(size=1,shape=16) + 
  theme_classic() + 
  scale_color_gradientn(colours = jdb_palette("solar_extra"),limits=c(-4,4),oob = scales::squish)+geom_text_repel(data = dat,aes(x=Corr.log10P,y=Enrichment.log10P,label=paste0("DORC: ",dat$DORC," Motif: ",dat$Motif)),colour = "black")

#only TFs
stim_FigR.tf<-stim_FigR.all.tf
dat <- subset(stim_FigR.tf, (2 < abs(Score)))

gAll.tf <- ggplot(stim_FigR.tf,aes(Corr.log10P,Enrichment.log10P,color=Score)) + 
  geom_point_rast(size=1,shape=16) + 
  theme_classic() + 
  scale_color_gradientn(colours = jdb_palette("solar_extra"),limits=c(-4,4),oob = scales::squish)+geom_text_repel(data = dat,aes(x=Corr.log10P,y=Enrichment.log10P,label=paste0("DORC: ",dat$DORC," Motif: ",dat$Motif)),color = "black")

stim_FigR.tf.wt<-stim_FigR.tf[stim_FigR.tf$DORC %in% WT_genes,]
dat <- subset(stim_FigR.tf.wt, (1 < abs(Score)))
gAll.tf.WT <- ggplot(stim_FigR.tf.wt,aes(Corr.log10P,Enrichment.log10P,color=Score)) + 
  geom_point_rast(size=1,shape=16) + 
  theme_classic() + 
  scale_color_gradientn(colours = jdb_palette("solar_extra"),limits=c(-4,4),oob = scales::squish)+geom_text_repel(data = dat,aes(x=Corr.log10P,y=Enrichment.log10P,label=paste0("DORC: ",dat$DORC," Motif: ",dat$Motif)),color = "black")

stim_FigR.tf.mut<-stim_FigR.tf[stim_FigR.tf$DORC %in% MUT_genes,]
dat <- subset(stim_FigR.tf.mut, (1 < abs(Score)))
gAll.tf.mut <- ggplot(stim_FigR.tf.mut,aes(Corr.log10P,Enrichment.log10P,color=Score)) + 
  geom_point_rast(size=1,shape=16) + 
  theme_classic() + 
  scale_color_gradientn(colours = jdb_palette("solar_extra"),limits=c(-4,4),oob = scales::squish)+geom_text_repel(data = dat,aes(x=Corr.log10P,y=Enrichment.log10P,label=paste0("DORC: ",dat$DORC," Motif: ",dat$Motif)),color = "black")

stim_FigR.tf.ctr<-stim_FigR.tf[stim_FigR.tf$DORC %in% CTR_genes,]
dat <- subset(stim_FigR.tf.ctr, (1 < abs(Score)))
gAll.tf.ctr <- ggplot(stim_FigR.tf.ctr,aes(Corr.log10P,Enrichment.log10P,color=Score)) + 
  geom_point_rast(size=1,shape=16) + 
  theme_classic() + 
  scale_color_gradientn(colours = jdb_palette("solar_extra"),limits=c(-4,4),oob = scales::squish)+geom_text_repel(data = dat,aes(x=Corr.log10P,y=Enrichment.log10P,label=paste0("DORC: ",dat$DORC," Motif: ",dat$Motif)),color = "black")



#only de

dat <- subset(stim_FigR_de, (-0.4 > Corr.log10P & Enrichment.log10P > 10 ) )

gAll_de <- ggplot(stim_FigR_de,aes(Corr.log10P,Enrichment.log10P)) + 
  geom_point_rast(size=1,shape=16) + 
  theme_classic() + 
  scale_color_gradientn(colours = jdb_palette("solar_extra"),limits=c(-4,4),oob = scales::squish)+geom_text_repel(data = dat,aes(x=Corr.log10P,y=Enrichment.log10P,label=paste0("DORC: ",dat$DORC," Motif: ",dat$Motif)))


#only mutant run DORCs
dat_MUT<-subset(stim_FigR_MUT, (-1 > Corr.log10P & Enrichment.log10P > 5 ) | (1 < Corr.log10P & Enrichment.log10P > 10 ) )
gAll_MUT <- ggplot(stim_FigR_MUT,aes(Corr.log10P,Enrichment.log10P,color=Score)) + 
  geom_point_rast(size=1,shape=16) + 
  theme_classic() + scale_color_gradientn(colours = jdb_palette("solar_extra"),limits=c(-4,4),oob = scales::squish)+geom_text_repel(data = dat_MUT,aes(x=Corr.log10P,y=Enrichment.log10P,label=paste0("DORC: ",DORC," Motif: ",Motif)),max.overlaps = 20)+ggtitle("MUT all DORCs")



#only WT run DORCs
dat_WT<-subset(stim_FigR_WT, (-1 > Corr.log10P & Enrichment.log10P > 5 ) | (1 < Corr.log10P & Enrichment.log10P > 10 ) )
gAll_WT <- ggplot(stim_FigR_WT,aes(Corr.log10P,Enrichment.log10P,color=Score)) + 
  geom_point_rast(size=1,shape=16) + 
  theme_classic() + scale_color_gradientn(colours = jdb_palette("solar_extra"),limits=c(-4,4),oob = scales::squish)+geom_text_repel(data = dat_WT,aes(x=Corr.log10P,y=Enrichment.log10P,label=paste0("DORC: ",DORC," Motif: ",Motif)),max.overlaps = 20)+ggtitle("WT all DORCs")

write.table(stim_FigR_WT,quote = F,row.names = F,sep = "\t",col.names = T,file="./IDH_vs_mutant/Only_WT_cells_all_Dorcs_TF_reg.txt")
write.table(stim_FigR_MUT,quote = F,row.names = F,sep = "\t",col.names = T,file = "./IDH_vs_mutant/Only_MUT_cells_all_Dorcs_TF_reg.txt")

dat_MUT <- subset(stim_FigR_de_MUT, (-0.5 > Corr.log10P & Enrichment.log10P > 5 ) | (1 < Corr.log10P & Enrichment.log10P > 2 ) )


score_MUT<-subset(stim_FigR_de_MUT, (Score > 1 ) )

write.csv(score_MUT,"./Top_MUT_TF_DORC_Corr.csv")

gAll_de_mut <- ggplot(stim_FigR_de_MUT,aes(Corr.log10P,Enrichment.log10P,color=Score)) + 
  geom_point_rast(size=1,shape=16) + theme_classic() + scale_color_gradientn(colours = jdb_palette("solar_extra"),limits=c(-4,4),oob = scales::squish)

gAll_de_mut<-gAll_de_mut+geom_text_repel(data = dat_MUT,aes(x=Corr.log10P,y=Enrichment.log10P,label=paste0("DORC: ",dat_MUT$DORC," Motif: ",dat_MUT$Motif)),color="black")+ggtitle("MUT associated TFs")


dat_WT <- subset(stim_FigR_de_WT, (1 < Corr.log10P & Enrichment.log10P > 10 ) )

score_WT<-subset(stim_FigR_de_WT, (0 < Corr.log10P & Score > 1 ) )

write.csv(score_WT,"./Top_WT_TF_DORC_Corr.csv")

gAll_de_WT <- ggplot(stim_FigR_de_WT,aes(Corr.log10P,Enrichment.log10P,color=Score)) + geom_point_rast(size=1,shape=16) + theme_classic() + scale_color_gradientn(colours = jdb_palette("solar_extra"),limits=c(-4,4),oob = scales::squish)

gAll_de_WT<-gAll_de_WT+geom_text_repel(data = dat_WT,aes(x=Corr.log10P,y=Enrichment.log10P,label=paste0("DORC: ",DORC," Motif: ",Motif)),colour = "black")+ggtitle("WT associated TFs")
gAll_de_mut+gAll_de_WT

# Mean plot
library(ggrepel)
rankDrivers(figR.d = stim_FigR) 

Idents(unintegrated_b1_b2_RNA)<-unintegrated_b1_b2_RNA$type
pos_markers<-FindMarkers(unintegrated_b1_b2_RNA,ident.1 = "MUT",only.pos = T)

pos_markers[row.names(pos_markers) %in% dorcGenes_WT,]

# Scatter plot of drivers associated with a given DORC
plotDrivers(figR.d = stim_FigR,marker = "IPCEF1")
plotDrivers(figR.d = stim_FigR,marker = "P2RY12")
plotDrivers(figR.d = stim_FigR,marker = "NAV3")

# Load DORC-implicated GWAS SNP P mat
SNPPmat <- readRDS("./data/GWAS/GWAS_DORC_SNPoV_unfiltered_P_mat.rds")
SNPDORCs <- rownames(SNPPmat)  # DORCs of interest
length(SNPDORCs)

# Heatmap 
figR_heat <- plotfigRHeatmap(figR.d = stim_FigR_de_WT,
                             score.cut = 1.5,
                             DORCs = SNPDORCs,
                             column_names_gp=gpar(fontsize=5),
                             show_row_dend = FALSE,
                             row_names_side = "left")

# D3Network
plotfigRNetwork(stim_FigR_MUT,score.cut = 1.5,DORCs = "DNAJB1")




```

Examples of drivers by using the faux paired dataset.
```{r}
library(motifmatchr)
library(GenomicRanges)
library(SummarizedExperiment)
library(TFBSTools)
library(JASPAR2020)
library(BSgenome.Hsapiens.UCSC.hg38)
pfm <- getMatrixSet(
  x = JASPAR2020,
  opts = list(collection = "CORE", species= 9606 , tax_group = 'vertebrates', all_versions = FALSE)
)
# Get motif matches for example motifs in peaks



Paired_pure_rename <- AddMotifs(Paired_pure_rename, genome = BSgenome.Hsapiens.UCSC.hg38, pfm = pfm)


motifMatches(motif_ix) # Extract matches matrix from SummarizedExperiment result

# Get motif positions within peaks for example motifs in peaks 
motif_ix <- matchMotifs(example_motifs, peaks, genome = "hg19",
                         out = "positions") 

saveRDS(object = Paired_pure_rename,file = "./Faux_multi_unique.rds")
Paired_pure_rename<-readRDS("./Faux_multi_unique.rds")



FindRegion <- function(
  object,
  region,
  sep = c("-", "-"),
  assay = NULL,
  extend.upstream = 0,
  extend.downstream = 0
) {
  if (!is(object = region, class2 = "GRanges")) {
    # first try to convert to coordinates, if not lookup gene
    region <- tryCatch(
      expr = suppressWarnings(
        expr = StringToGRanges(regions = region, sep = sep)
      ),
      error = function(x) {
        region <- LookupGeneCoords(
          object = object,
          assay = assay,
          gene = region
        )
        return(region)
      }
    )
    if (is.null(x = region)) {
      stop("Gene not found")
    }
  }
  region <- suppressWarnings(expr = Extend(
    x = region,
    upstream = extend.upstream,
    downstream = extend.downstream
  )
  )
  return(region)
}







plot_reg<-function(object,gene,TFs,extend.upstream=10000,extend.downstream=10000,plot_expr=NULL){
#gene_region<-regionlookup(object=object,region = gene,extend.upstream = extend.upstream,extend.downstream = extend.downstream)
  
gene_region<-FindRegion(object=object,region = gene,extend.upstream = extend.upstream,extend.downstream = extend.downstream)
  
  
peaks<-StringToGRanges(row.names(object))
select_peaks<-peaks[queryHits(findOverlaps(peaks,gene_region))]
# Get motif matches for example motifs in peaks
motif_ix <- matchMotifs(pfm, select_peaks, genome = "hg38") 
mot<-GetMotifData(object = object, assay = "ATAC", slot = "motif.names")
mot2<-with(stack(mot), split(as.character(ind), values))
sel<-unlist(mot2[TFs])

peak_highlight<-GRanges(seqnames=NULL,ranges=NULL,strand=NULL, binding=NULL)
for (i in 1:length(sel)) {
tf_site<-select_peaks[assays(motif_ix)$motifMatches[,sel[i]]]
tf_site$binding<-names(sel)[i]

peak_highlight<-c(peak_highlight,tf_site)
}

test<-BigwigTrack(region = gene_region,bigwig = list(H3K27="./CNR/CnR_H3K27ac_0125.hg38.sorted.RmDup.10mNorm.bw"),type = 'heatmap')


if(is.null(plot_expr)){
p<-CoveragePlot(Paired_pure_rename,region = gene_region,expression.assay = "RNA",features = TFs,links = T,ranges = peak_highlight,region.highlight = peak_highlight,extend.upstream = extend.upstream,extend.downstream = extend.downstream, ranges.group.by = "binding",cols = small_color_map,bigwig = list(H3K27="./CNR/CnR_H3K27ac_0125.hg38.sorted.RmDup.10mNorm.bw") 
)  

} else {
p<-CoveragePlot(Paired_pure_rename,region = gene_region,expression.assay = "RNA",features = plot_expr,links = T,ranges = peak_highlight,region.highlight = peak_highlight,extend.upstream = extend.upstream,extend.downstream = extend.downstream, ranges.group.by = "binding",cols = small_color_map,bigwig = list(H3K27="./CNR/CnR_H3K27ac_0125.hg38.sorted.RmDup.10mNorm.bw")) 
}


return(p)
}


conservation_mm10<-BigwigTrack(region = reg_m,bigwig = list(H3K27="./CNR/CnR_H3K27ac_0125.hg38.sorted.RmDup.10mNorm.bw"),type = 'heatmap')



plot_reg(Paired_pure_rename,gene = "SYNDIG1",TFs  = c("MEF2C","FOSL2"))

BigwigTrack(region = LookupGeneCoords(gene = "CD163",object = Paired_pure_rename),bigwig = list(H3K27="./CNR/CnR_H3K27ac_0125.hg38.sorted.RmDup.10mNorm.bw"),type = 'heatmap')


plot_reg(Paired_pure_rename,gene = "CD163",TFs  = c("FOSL2"),extend.upstream = 50000,extend.downstream = 50000,plot_expr = c("CD163","FOSL2"))

plot_reg(Paired_pure_rename,gene = "CD163",TFs  = c("MEF2C","FOSL2"),extend.upstream = 50000,extend.downstream = 50000)

plot_reg(Paired_pure_rename,gene = "CD74",TFs  = c("CEBPD"),extend.upstream = 50000,extend.downstream = 50000)

plot_reg(Paired_pure_rename,gene = "CD163",TFs  = c("FOSL2"),extend.upstream = 50000,extend.downstream = 50000,plot_expr = c("CD163","FOSL2"))

#IQGAP1, STAB1, CD44, ANXA1, ITGAM, CD163 and S100A6?

p1<-plot_reg(Paired_pure_rename,gene = "IQGAP1",TFs  = c("FOSL2"),extend.upstream = 50000,extend.downstream = 50000,plot_expr = c("IQGAP1","FOSL2"))
p2<-plot_reg(Paired_pure_rename,gene = "STAB1",TFs  = c("FOSL2"),extend.upstream = 50000,extend.downstream = 50000,plot_expr = c("STAB1","FOSL2"))
p3<-plot_reg(Paired_pure_rename,gene = "CD44",TFs  = c("FOSL2"),extend.upstream = 50000,extend.downstream = 50000,plot_expr = c("CD44","FOSL2"))
p4<-plot_reg(Paired_pure_rename,gene = "ANXA1",TFs  = c("FOSL2"),extend.upstream = 50000,extend.downstream = 50000,plot_expr = c("ANXA1","FOSL2"))
p5<-plot_reg(Paired_pure_rename,gene = "ITGAM",TFs  = c("FOSL2"),extend.upstream = 50000,extend.downstream = 50000,plot_expr = c("ITGAM","FOSL2"))
p6<-plot_reg(Paired_pure_rename,gene = "CD163",TFs  = c("FOSL2"),extend.upstream = 50000,extend.downstream = 50000,plot_expr = c("CD163","FOSL2"))
```

